vdr/ambassador.go

```diff
@@ -37,20 +49,245 @@ type Ambassador interface {
+	// Resolve controllers of current version (could be the same document)
+	didControllers, err := n.resolveDIDControllers(currentDIDDocument)
+	logging.Log().Debug(didControllers)
+
+	var controllerVerificationRelationships = []did.VerificationRelationship{}
+	for _, didCtrl := range didControllers {
+		for _, auth := range didCtrl.Authentication {
+			controllerVerificationRelationships = append(controllerVerificationRelationships, auth)
+		}
+	}
+
+	// In an update, only the keyID is provided in te network document. Resolve the key from the key store
+	// This should succeed since the signature of the network document has already been verified.
+	pKey, err := n.keyResolver.GetPublicKey(document.SigningKeyID(), document.SigningTime())
+	if err != nil {
+		return fmt.Errorf("unable to resolve signingkey %w", err)
+	}
+	headerKey, err := jwk.New(pKey)
+	if err != nil {
+		return fmt.Errorf("could not parse public key into jwk %w", err)
+	}
+	// Create thumbprint
+	headerKeyThumbprint, err := headerKey.Thumbprint(thumbprintAlg)
+	if err != nil {
+		return fmt.Errorf("unable to generate network document signing key thumbprint")
+	}
+
+	// Check if the signingKey is listed as a valid authenticationMethod in one of the controllers
+	keyToSign, err := n.findKeyByThumbprint(headerKeyThumbprint, controllerVerificationRelationships)
+	if keyToSign == nil {
+		return fmt.Errorf("network document not signed by one of its controllers")
+	}
+
+	// TODO: perform all these tests:
+	// Take authenticationMethod keys from the controllers
+	// Check if network header keyID is one of authenticationMethods of the controller
+	//
+	// For each verificationMethod in the next version document
+	// 		check if the provided key thumbprint matches the corresponding thumbprint in the key store
+	// Take diff of verificationMethods between next and current versions:
+	// if new verificationMethod is added:
+	// 		Add public key to key store
+	// if verificationMethod is removed:
+	//		Mark keyID as expired since the updatedAt time from new DID document
+
+	// make a diff of the controllers
+	// 	if controller is added
+	//		check if it is known.
+	updatedAt := document.SigningTime()
+	documentMetadata := types.DocumentMetadata{
+		Created:    currentDIDMeta.Created,
+		Updated:    &updatedAt,
+		Version:    document.TimelineVersion(),
+		TimelineID: document.TimelineID(),
+		Hash:       document.PayloadHash(),
+	}
+	return n.didStore.Update(nextDIDDocument.ID, currentDIDMeta.Hash, nextDIDDocument, &documentMetadata)
+}
+
+// checkSubscriberDocumentIntegrity performs basic integrity checks on the SubscriberDocument fields:
+func checkSubscriberDocumentIntegrity(document dag.SubscriberDocument) error {
+	// check the payload type:
+	if document.PayloadType() != DIDDocumentType {
+		return fmt.Errorf("wrong payload type for this subscriber. Can handle: %s, got: %s", DIDDocumentType, document.PayloadType())
+	}
+
+	// PayloadHash must be set
+	if document.PayloadHash().Empty() {
```

Comment:
this can't happen in practice, if it's something we'd like to check it should be in the network layer (not here). Your subscribers would also never be called when there's no payload (since it's called after payload is received).

Reply:
So I can convert it to a panic instead of error? :)